### DISCUSSION TITLE
Can't run zephyr application from SDRAM on RT1060-EVKB

### DISCUSSION THREAD

**Discussion by ofirshe** [zephyrproject-rtos/zephyr#48597] [Q&A]
Hey, 

I wanted to link my application to SDRAM and then download the zephyr.elf using the debugger for debug purposes (can't debug my application from the FLASH). 
I tried to run my zephyr application from the SDRAM using CONFIG_XIP=n and CONFIG_NO_OPTIMIZATIONS=y, but the application hangs in arm_mpu_init(). In order to overcome this, I set CONFIG_ARM_MPU=n but I get warning regarding stack protection. 
When will execution from SDRAM be supported by zephyr ? 

Best Regards, 
Ofir.

**Comment by raghulshanmuganathan** [comment]
I can run zephyr.elf from a debugger by removing the following lines from Kconfig.defconfig in my custom board directory (MIMXRT1060).

choice CODE_LOCATION
	default CODE_FLEXSPI
endchoice

It defaults to CODE_ITCM for me if I remove it. You can try the same maybe!

**Reply by ofirshe** [reply]
it doesn't work for me unfortunately.  I can't enter to main function because it hangs in z_sys_init_run_level(_SYS_INIT_LEVEL_PRE_KERNEL_1);.
I did the same as you. Can you please have a look on my custom board ? Thanks.
[boards.zip](https://github.com/zephyrproject-rtos/zephyr/files/9249463/boards.zip)

**Reply by ofirshe** [reply]
the debugger jumps to fatal error

**Reply by ofirshe** [reply]
this is the entry cause to this bug 0x00004870        0x8 zephyr/kernel/libkernel.a(kheap.c.obj)

**Reply by ofirshe** [reply]
That's the output from the debuuger: 
loadfile "zephyr.elf"
Loaded section .ARM.exidx: 0x0000472C ~ 0x00004733 (size 0x8)
Loaded section rom_start: 0x00000000 ~ 0x000002BF (size 0x2C0)
Loaded section text: 0x000002C0 ~ 0x0000472B (size 0x446C)
Loaded section .ARM.exidx: 0x0000472C ~ 0x00004733 (size 0x8)
Loaded section initlevel: 0x00004734 ~ 0x000047CB (size 0x98)
Loaded section devices: 0x000047CC ~ 0x00004903 (size 0x138)
Loaded section sw_isr_table: 0x00004904 ~ 0x00004E03 (size 0x500)
Loaded section device_handles: 0x00004E04 ~ 0x00004E5D (size 0x5A)
Loaded section ctors: 0x00004E5E ~ 0x00004E6B (size 0xE)
Loaded section rodata: 0x00004E6C ~ 0x0000662B (size 0x17C0)
Loaded section datas: 0x20000000 ~ 0x20000137 (size 0x138)
Loaded section device_states: 0x20000138 ~ 0x2000016B (size 0x34)
Loaded section k_heap_area: 0x2000016C ~ 0x20000183 (size 0x18)
Entry point 0x00001501

**Reply by ofirshe** [reply]
and if you can send me your custom board folder I'll be grateful.

**Reply by DerekSnell** [reply]
Hi @ofirshe ,
When I changed the Kconfig symbols to execute from ITCM, but left the data in SDRAM, I had faults when the SDRAM was accessed.  This was caused when I used the debugger to load and start the application rather than booting from flash.  [This discussion](https://github.com/zephyrproject-rtos/zephyr/discussions/36800#discussioncomment-3261143) explains the reason for this, and options to workaround it.

**Reply by ofirshe** [reply]
Hey @DerekSnell, 

I understand the problem that when my application go through the RAM, it skips all the SDRAM initialization which occurs from the Boot Rom (DCD). Thus, according to your response, I must concatenate somehow the MCUBoot at the start to my application.

I have many questions:
1. How can I build the MCUBoot with my application in a single elf file ? 
2. After generating that elf, will I be possible to upload that from the debugger ? Because I believe it includes some writes to FLASH and it's Read Only Memory (might get bus error). 

what I've done so far in my custom board folder:
1. Change Code location to CODE_ITCM in Kconfig.defconfig
2. Change sram location in dts file - changed 'zephyr,sram = &sdram0;' to 'zephyr,sram = &dtcm;'


Can you please write down here all the steps I have to do to compile my application with the MCUBoot included in ? I'll appreciate that a lot. 

Or maybe is there another shorter workaround which I can do ? 

Thanks you very much,
Ofir.

**Reply by DerekSnell** [reply]
Hi @ofirshe ,
I shared the detailed steps for this at [this other discussion](https://github.com/zephyrproject-rtos/zephyr/discussions/36800#discussioncomment-3325663), because that discussion already has many details about using the TCMs, and the considerations when booting.

Please give this a try, and let us know how it works for you.
